utils/swift-rpathize.py

```diff
@@ -0,0 +1,86 @@
+# paths beginning with `@executable_path` or `@loader_path`, meaning the
+# path containing the running executable or the image being loaded,
+# respectively.
+#
+# Many of Swift's dylibs are meant to be installed on the system, which
+# means they have install names like this:
+#   /usr/lib/swift/libswiftFoo.dylib
+# To support back-deployment, they also provide magic override symbols
+# ($ld$install_name) for all the OS versions preceding the addition of
+# of the library.  When the linker finds a dylib with a matching override
+# for the OS deployment target, it ignores the normal install name and
+# uses the override path in the linked image's load command.  Swift's
+# libraries use override paths that begin with `@rpath`, and Swift
+# builds images with a runtime search path list that starts with
+# /usr/lib/swift but then falls back on a path relative to the image;
+# thus, apps will use the system libraries if available but will
+# otherwise use fallback libraries.
+#
+# When we're working on Swift, we usually want to test the libraries
+# we just built rather than the system libraries.  There are two ways
+# to achieve that.  The first is to override dyld's runtime search path
+# with DYLD_LIBRARY_PATH; this will take precedence over even an
+# absolute install name.  The second is to make sure the dylibs are
+# loaded via an @rpath install name and then link the program with an
+# rpath that will use the just-built libraries.  Unfortunately, the
+# toolchain will ordinarily use an absolute install name instead of
+# an @rpath if the deployment target is old enough, subverting testing.
+#
+# This script looks for dependent dylibs with an absolute path in
+# /usr/lib/swift and changes them to use @rpath.
+
+import argparse
+import re
+import subprocess
+import sys
+
+
+def main(arguments):
+    parser = argparse.ArgumentParser(
+        description='Change absolute install names to use @rpath')
+    parser.add_argument('bin', help='the binary')
+
+    args = parser.parse_args(arguments)
+    rpathize(args.bin)
+
+
+def rpathize(filename):
+    dylibsOutput = subprocess.check_output(
+        ['xcrun', 'dyldinfo', '-dylibs', filename])
+
+    # The output from dyldinfo -dylibs is a line of header followed by one
+    # install name per line, indented with spaces.
+    dylib_regex = re.compile(
+        r"^\s*(?P<path>/usr/lib/swift/(?P<filename>.*\.dylib))\s*$")
+
+    # Build a command to invoke install_name_tool.
+    command = ['install_name_tool']
+    for line in dylibsOutput.splitlines():
+        match = dylib_regex.match(line)
+        if match:
+            command.append('-change')
+            command.append(match.group('path'))
+            command.append('@rpath/' + match.group('filename'))
+            continue
+
+    # Don't run the command if we didn't find any dylibs to change:
+    # it's invalid to invoke install_name_tool without any operations.
+    if len(command) == 1:
```

Comment:
Are there legitimate cases where this would happen? It would be nice to check for whatever those cases are and fail otherwise, so some change in output from the tool doesn't make us go back to silently testing the OS runtime.

Reply:
The example that hit this during PR testing was just a unit test that didn't link any Swift runtime libraries at all — I think it was testing something in Sema.  But I think in principle you could also get it on the right deployment target, where the libraries declare an `@rpath` override for their install name.